ArduPilotDriver.py

```diff
@@ -86,54 +88,53 @@ def cleanup():
-                last_data_send_times["orientation"] = curr_time
-                asyncio.run(send_orientation_message(websocket, attitude.roll, attitude.pitch, attitude.yaw))
-        
-        def heading_callback(self, _, heading):
-            curr_time = time.perf_counter()
-            if curr_time >= last_data_send_times["heading"] + 1./args.frequency:
-                if args.debug:
-                    print("Heading", heading)
-                last_data_send_times["heading"] = curr_time
-                asyncio.run(send_heading_message(websocket, heading))
-        
-        ardupilot.add_attribute_listener("location.global_frame", gps_callback)
-        ardupilot.add_attribute_listener("attitude", orientation_callback)
-        # ardupilot.add_attribute_listener("heading", heading_callback)
-
+    while True:
         try:
-            while True:
-                await asyncio.sleep(1)
-        except websockets.ConnectionClosed as e:
-            print(e)
-            ardupilot.remove_attribute_listener("location.global_frame", gps_callback)
-            ardupilot.remove_attribute_listener("attitude", orientation_callback)
-            # ardupilot.remove_attribute_listener("heading", heading_callback)
-            continue
+            async with websockets.connect("ws://localhost:3001/ardupilot") as websocket:
+                print("Done!")
+
+                def gps_callback(self, _, frame):
+                    curr_time = time.perf_counter()
+                    if curr_time >= last_data_send_times["gps"] + 1./args.frequency:
+                        if args.debug:
+                            print("GPS Fix:", ardupilot.gps_0.fix_type, frame)
+                        if ardupilot.gps_0.fix_type > 1:
+                            last_data_send_times["gps"] = curr_time
+                            asyncio.run(send_gps_message(websocket, frame.lat, frame.lon))
+
+                def orientation_callback(self, _, attitude):
+                    curr_time = time.perf_counter()
+                    if curr_time >= last_data_send_times["orientation"] + 1./args.frequency:
+                        if args.debug:
+                            print(attitude)
+                        else:
+                            pitch_deg = attitude.pitch * 180 / math.pi
+                            roll_deg = attitude.roll * 180 / math.pi
+                            print(f"\33[2K\rroll={roll_deg: 4.0f}, pitch={pitch_deg: 4.0f}", end="")
+                        last_data_send_times["orientation"] = curr_time
+                        asyncio.run(send_orientation_message(
+                            websocket, attitude.roll, attitude.pitch, attitude.yaw))
+
+                ardupilot.add_attribute_listener("location.global_frame", gps_callback)
+                ardupilot.add_attribute_listener("attitude", orientation_callback)
+
+                try:
+                    asyncio.get_event_loop().run_forever()
```

Comment:
It looks like `get_event_loop` is [deprecated](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.get_event_loop). Is there another way to do this? Not a big deal if not, just don't want it to start throwing errors all of a sudden.

Reply:
Could you use `get_running_loop`? Looks like it would require a bit more work in terms of error checking and also not super clear on the difference between the two.